### PR TITLE
[DMs] Websockets middleware for Chats

### PR DESCRIPTION
**Pull request by rickyrombo**
### Description

Creates a middleware that listens for chat messages and reactions and updates state accordingly.

Depends on either:
- https://github.com/AudiusProject/audius-protocol/pull/4723
- https://github.com/AudiusProject/audius-protocol/pull/4704

Probably needs a little adjustment to encapsulate the event emitter better. But that's the gist.

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

### CODE REVIEW

File: packages/web/src/pages/chat-page/components/ChatList.tsx [hunk 1]
```diff
@@ -76,7 +76,9 @@ export const ChatList = (props: ChatListProps) => {
             <div className={styles.header}>{messages.nothingHere}</div>
             <div className={styles.subheader}>{messages.start}</div>
           </div>
-        ) : null}
+        ) : (
+          <LoadingSpinner className={styles.spinner} />
```
Comment (rickyrombo) [inline packages/web/src/pages/chat-page/components/ChatList.tsx:80]: TODO: error state here

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/mjp-dms-websocket-middleware

File: packages/common/src/store/pages/chat/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+              setMessageReactionSucceeded({
+                chatId,
+                messageId,
+                reaction
+              })
+            )
+          }
+          closeListener = () => {
+            console.log('Chat middleware WebSocket closed. Reconnecting...')
+            sdk.chats.listen()
```
Comment (rickyrombo) [inline packages/common/src/store/pages/chat/middleware.ts:46]: Might need some more clever logic here to not spam forever when a DN goes down or becomes unreachable, and eventually reselect...

File: packages/common/src/store/pages/chat/middleware.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  (store) => {
+    let messageListener: ChatEvents['message'] | null = null
+    let reactionListener: ChatEvents['reaction'] | null = null
+    let openListener: ChatEvents['open'] | null = null
+    let closeListener: ChatEvents['close'] | null = null
+
+    return (next) => (action) => {
+      if (connect.match(action)) {
+        console.log('Chat middleware attaching')
+        audiusSdk().then((sdk) => {
```
Comment (piazzatron) [inline packages/common/src/store/pages/chat/middleware.ts:25]: nit: any reason using promises vs async await?

File: packages/common/src/store/pages/chat/selectors.ts [hunk 4]
```diff
@@ -11,12 +10,15 @@ const { getUserId } = accountSelectors
...
 
 export const getChat = (state: CommonState, chatId?: string) =>
-  chatId
-    ? state.pages.chat.chatList.data?.find((chat) => chat.chat_id === chatId)
-    : undefined
+  chatId ? state.pages.chat.chatList.map[chatId] : undefined
 
-export const getChatsRaw = (state: CommonState) =>
-  state.pages.chat.chatList.data
+export const getChatsRaw = createSelector(
```
Comment (piazzatron) [inline packages/common/src/store/pages/chat/selectors.ts:15]: niceee

File: packages/common/src/store/pages/chat/slice.ts [hunk 5]
```diff
@@ -9,7 +15,8 @@ type ChatState = {
   chatList: {
     status: Status
     summary?: TypedCommsResponse<UserChat>['summary']
-    data: UserChat[]
+    map: Record<string, UserChat>
```
Comment (piazzatron) [inline packages/common/src/store/pages/chat/slice.ts:18]: nit `chatIdMap`?

File: packages/common/src/store/pages/chat/middleware.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+export const chatMiddleware =
+  (audiusSdk: () => Promise<ReturnType<typeof sdk>>): Middleware =>
+  (store) => {
+    let messageListener: ChatEvents['message'] | null = null
+    let reactionListener: ChatEvents['reaction'] | null = null
+    let openListener: ChatEvents['open'] | null = null
+    let closeListener: ChatEvents['close'] | null = null
+
+    return (next) => (action) => {
+      if (connect.match(action)) {
```
Comment (piazzatron) [inline packages/common/src/store/pages/chat/middleware.ts:23]: Wonder if we should track some global connection state? i.e. if we hit two connect actions we're going to register double listeners, seems like a bug waiting to happen...

**Review by piazzatron — APPROVED** [review]
SWEET! 

For my own curiosity, why did we opt for a middleware approach here vs our more typical saga convention?

File: packages/common/src/store/pages/chat/middleware.ts (hunk 3 above)
Comment (rickyrombo) [reply packages/common/src/store/pages/chat/middleware.ts:25]: I suppose I could wrap the contents in a callable async fn. I don't think the middleware should be async, though.

File: packages/common/src/store/pages/chat/middleware.ts (hunk 6 above)
Comment (rickyrombo) [reply packages/common/src/store/pages/chat/middleware.ts:23]: Yeah good idea. I originally had something for this back when the chatsApi returned the websocket but forgot to add something back to track

**Comment by rickyrombo** [comment]
> SWEET!
> 
> For my own curiosity, why did we opt for a middleware approach here vs our more typical saga convention?

Middleware was easier to wrap my head around - I tend to think of sagas as very linear (even if async), not as event based.

**Pull request merged**